hawc/apps/epi/api.py

```diff
@@ -87,37 +99,410 @@ def result_heatmap(self, request, pk):
+                            raf_ids_to_disassociate.append(existing_result_af.id)
+
+                # Now - af_ids_to_create_for_type either contains all the ids (if during a create)
+                # or just the ones that weren't removed b/c they are already in the db. We can now
+                # build up just the inserts that actually need to happen.
+                for adjustment_factor_id in af_ids_to_create_for_type:
+                    dynamic_obj = {
+                        "included_in_final_model": is_included,
+                        "adjustment_factor": adjustment_factor_id,
+                        "result": result_id,
+                    }
+                    inserts.append(dynamic_obj)
+
+        # Now that we've looked at each category we can hit the db.
 
-class ComparisonSet(AssessmentViewset):
+        # Wipe out existing criteria for each ResultAdjustmentFactor.id that acutally needs deleting
+        if len(raf_ids_to_disassociate) > 0:
+            models.ResultAdjustmentFactor.objects.filter(id__in=raf_ids_to_disassociate).delete()
+
+        # ...and save any new ones
+        if len(inserts) > 0:
+            serializer = serializers.SimpleResultAdjustmentFactorSerializer(data=inserts, many=True)
+            serializer.is_valid(raise_exception=True)
+            serializer.save()
+
+    def handle_adjustment_factors(self, result):
+        """
+        Converts adjustment factor input to id's; creating if necessary.
+
+        Args:
+            result (Result): the just created result
+        """
+        assessment_id = result.get_assessment().id
+
+        # Now do some conversions/creations
+        for fc in self.factor_categories:
+            data_key = fc[0]
+
+            # Client can supply an id, or the name of the factor entry (and then we'll look it up for them - or create it if needed)
+            if data_key in self.request.data:
+                data_probe = self.request.data[data_key]
+
+                fixed = []
+                for el in data_probe:
+                    if type(el) is str:
+                        try:
+                            adj_factor = models.AdjustmentFactor.objects.get(
+                                description=el, assessment_id=assessment_id
+                            )
+                            fixed.append(adj_factor.id)
+                        except ObjectDoesNotExist:
+                            # Allow creation of adjustment factors as part of the request
+                            af_serializer = serializers.AdjustmentFactorSerializer(
+                                data={"description": el, "assessment": assessment_id}
+                            )
+                            af_serializer.is_valid(raise_exception=True)
+                            af = af_serializer.save()
+                            fixed.append(af.id)
+                    else:
```

Comment:
any type checking here, or anythign goes as long as it's not a str?

Reply:
if it's anything other than a string, you'll get normal django checks/errors, for instance if you pass in a bool you'll get:

`[`
`  {`
`    "adjustment_factor": [`
`      "Invalid pk \"True\" - object does not exist."`
`    ]`
`  }`
`]`